lib/link_checker.rb

```diff
@@ -0,0 +1,49 @@
+require "uri"
+
+module LinkChecker
+module_function
+
+  def gather_urls(yaml_node)
+    # Gather urls in a Set to allow simple combination and duplicate removal
+    urls = Set[]
+    case yaml_node
+    when String
+      urls.merge(Set.new(URI.extract(yaml_node, %w[http https]).to_set))
+    when Hash
+      yaml_node.each_value do |value|
+        urls.merge(gather_urls(value))
+      end
+    when Array
+      yaml_node.each do |value|
+        urls.merge(gather_urls(value))
+      end
+    end
+    urls
+  end
+
+  def get_urls_from_locale_files
+    # For each file, gather the urls as a set and then combine the sets together using reduce
+    Dir.glob("config/locales/*.yml")
+     .map { |locale_filename| gather_urls YAML.safe_load(File.read(locale_filename)) }
+     .reduce(:+)
```

Comment:
Combining the gathering of data from YAML files with extracting the URLs from the strings in one class level method, make this very difficult to test. Hence the spec for this method is of full stubbing. IMO This should be broken down into two methods. One to gather the raw data from the YAML files, and another to extract the urls from that data. Then each part of the process can be tested in isolation and without stubbing.